extendr-api/src/ownership.rs

```diff
@@ -0,0 +1,328 @@
+mod test {
+    use super::*;
+    use crate::*;
+    use libR_sys::{Rf_ScalarInteger, Rf_protect, Rf_unprotect};
+
+    #[test]
+    fn basic_test() {
+        test! {
+            single_threaded(|| unsafe {
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                }
+
+                let sexp1 = Rf_protect(Rf_ScalarInteger(1));
+                let sexp2 = Rf_protect(Rf_ScalarInteger(2));
+                protect(sexp1);
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                    assert_eq!(own.ref_count(sexp1), 1);
+                    assert_eq!(own.ref_count(sexp2), 0);
+                }
+
+                protect(sexp1);
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                    assert_eq!(own.ref_count(sexp1), 2);
+                    assert_eq!(own.ref_count(sexp2), 0);
+                }
+
+                unprotect(sexp1);
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                    assert_eq!(own.ref_count(sexp1), 1);
+                    assert_eq!(own.ref_count(sexp2), 0);
+                }
+
+                unprotect(sexp1);
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                    assert_eq!(own.ref_count(sexp1), 0);
+                    assert_eq!(own.ref_count(sexp2), 0);
+                }
+
+                protect(sexp2);
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                    assert_eq!(own.ref_count(sexp1), 0);
+                    assert_eq!(own.ref_count(sexp2), 1);
+                }
+
+                protect(sexp1);
+                {
+                    let mut own = OWNERSHIP.lock().expect("lock failed");
+                    own.check_objects();
+                    assert_eq!(own.ref_count(sexp1), 1);
+                    assert_eq!(own.ref_count(sexp2), 1);
+                }
+                Rf_unprotect(2);
+            });
+        }
+    }
+
+    #[test]
+    fn collection_test() {
+        test! {
+            single_threaded(|| unsafe {
+                {
+                    let mut own = OWNERSHIP.lock().expect("protect failed");
+                    own.check_objects();
+                }
+
+                // Force a garbage collect.
+                let test_size = INITIAL_PRESERVATION_SIZE + EXTRA_PRESERVATION_SIZE * 5;
+
+                // Make some test objects.
+                let sexps = (0..test_size).map(|_| {
+                    let sexp = Rf_ScalarInteger(1);
+                    R_PreserveObject(sexp);
```

Comment:
I bet here you're triggering the issue described here: https://github.com/RcppCore/Rcpp/issues/1081
The crash was fixed in recent versions of R, that's why only the oldrel CI runs fail.

Any reason you're using `R_PreserveObject()` instead of `Rf_protect()`? The latter shouldn't cause the problem.